std/experimental/allocator/building_blocks/bitmapped_block.d

```diff
@@ -229,571 +173,1462 @@ struct BitmappedBlock(size_t theBlockSize, uint theAlignment = platformAlignment
+        /*
+        Adjusts the memoized _startIdx to the leftmost control word that has at
+        least one zero bit. Assumes all control words to the left of $(D
+        _control[_startIdx]) are already occupied.
+        */
+        private void adjustStartIdx()
+        {
+            while (_startIdx < _control.rep.length
+                && _control.rep[_startIdx] == ulong.max)
+            {
+                static if (isShared)
+                {
+                    // Shared demands atomic increment, however this is protected
+                    // by a lock. Regular increment is fine
+                    auto localStart = _startIdx + 1;
+                    _startIdx = localStart;
+                }
+                else
+                {
+                    ++_startIdx;
+                }
+            }
+        }
+
+        /*
+        Based on the latest allocated bit, 'newBit', it adjusts '_freshBit'
+        */
+        private void adjustFreshBit(this _)(const ulong newBit)
+        {
+            import std.algorithm.comparison : max;
+            _freshBit = max(newBit, _freshBit);
+        }
+
+        /*
+        Returns the blocks corresponding to the control bits starting at word index
+        wordIdx and bit index msbIdx (MSB=0) for a total of howManyBlocks.
+        */
+        @trusted
+        private void[] blocksFor(this _)(size_t wordIdx, uint msbIdx, size_t howManyBlocks)
+        {
+            assert(msbIdx <= 63);
+            const start = (wordIdx * 64 + msbIdx) * blockSize;
+            const end = start + blockSize * howManyBlocks;
+            if (start == end) return null;
+            if (end <= _payload.length) return cast(void[]) _payload[start .. end];
+            // This could happen if we have more control bits than available memory.
+            // That's possible because the control bits are rounded up to fit in
+            // 64-bit words.
+            return null;
+        }
+
+        @trusted void[] allocate(const size_t s)
+        {
+            static if (isShared)
+            {
+                lock.lock();
+                scope(exit) lock.unlock();
+            }
+            return allocateImpl(s);
+        }
+
+        // If shared, this is protected by a lock inside 'allocate'
+        private @trusted void[] allocateImpl(const size_t s)
+        {
+            const blocks = s.divideRoundUp(blockSize);
+            void[] result = void;
+
+        switcharoo:
+            switch (blocks)
+            {
+            case 1:
+                // inline code here for speed
+                // find the next available block
+                foreach (i; _startIdx .. _control.rep.length)
+                {
+                    const w = _control.rep[i];
+                    if (w == ulong.max) continue;
+                    uint j = leadingOnes(w);
+                    assert(j < 64);
+                    assert((_control.rep[i] & ((1UL << 63) >> j)) == 0);
```

Comment:
`assert` should always have a message explaining why it failed. This will make it easier for people to read the code and in the future if the assert fails for people debugging it.